Just-in-time data mapping

davidscn opened the pull request:
## Main changes of this PR

Implements a `read-consistent` and `write-conservative` just-in-time mapping for nearest neighbor mapping and PU-RBF. For time interpolation we do (as discussed) a sample, then map.

We also change the behavior as follows:

- renamed the 'direct-access' flag to 'api-access'.
- The API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" require now the flag api access="true". If not set, the function will throw an error.
- When combining direct-mesh access with mappings, the function "getMeshVertexIDsAndCoordinates" will return only the vertices within the defined access region ("setMeshAccessRegion"). Setting an access region is mandatory.
- Added experimental support for an just-in-time mapping with NN and PU-RBF mapping via the new API functions 'mapAndReadData' and 'mapAndWriteData'.
- Still a bit open what to do with the combination of direct access with mappings. Right now we have: if nothing else is defined (only direct access) an access region needs to be defined (requirement from the partitioning). If we run in parallel, it's anyway required. If we combine a direct-access with a mapping, we don't need to define the access region. I am not sure if we want to keep the latter exception. Changing it would be breaking though.


* [x] Nearest-neighbor mapping (read and write)
* [x] PU-RBF mapping (read and write)
* [x] Testing: serial nearest-neighbor mapping (read and write)
* [x] Testing: serial PU-RBF mapping (read and write)
* [x] Parallization/repartitioning
* [x] Testing: parallel nearest-neighbor mapping (read)
* [x] Testing: parallel PU-RBF mapping (read and write)
* [x] More extensive parallel testing

## Motivation and additional information

See also #1843 and #2162


## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I stuck to C++17 features.
* [x] I stuck to CMake version 3.22.1.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

<!-- Tag people next to each point and add points for specific questions -->

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

<!-- add more questions/tasks if necessary -->

uekerman commented on src/precice/Participant.cpp:
```diff
@@ -215,6 +215,25 @@ void Participant::readData(
   _impl->readData(toSV(meshName), toSV(dataName), ids, relativeReadTime, values);
 }
 
+void Participant::mapAndreadData(
+    ::precice::string_view        meshName,
+    ::precice::string_view        dataName,
+    ::precice::span<const double> coordinates,
+    double                        relativeReadTime,
+    ::precice::span<double>       values) const
+{
+  _impl->mapAndreadData(toSV(meshName), toSV(dataName), coordinates, relativeReadTime, values);
+}
+
+void Participant::mapAndwriteData(
+    ::precice::string_view        meshName,
+    ::precice::string_view        dataName,
+    ::precice::span<const double> coordinates,
+    ::precice::span<const double> values)
+{
+  _impl->mapAndwriteData(toSV(meshName), toSV(dataName), coordinates, values);
+}
```
```suggestion
void Participant::mapAndReadData(
    ::precice::string_view        meshName,
    ::precice::string_view        dataName,
    ::precice::span<const double> coordinates,
    double                        relativeReadTime,
    ::precice::span<double>       values) const
{
  _impl->mapAndReadData(toSV(meshName), toSV(dataName), coordinates, relativeReadTime, values);
}

void Participant::mapAndWriteData(
    ::precice::string_view        meshName,
    ::precice::string_view        dataName,
    ::precice::span<const double> coordinates,
    ::precice::span<const double> values)
{
  _impl->mapAndWriteData(toSV(meshName), toSV(dataName), coordinates, values);
}
```

davidscn commented on src/mesh/config/MeshConfiguration.cpp:
```diff
@@ -99,6 +99,19 @@ const PtrDataConfiguration &MeshConfiguration::getDataConfiguration() const
   return _dataConfig;
 }
 
+mesh::PtrMesh MeshConfiguration::getIndirectAccessMesh(int dimension)
+{
+  PRECICE_ASSERT(dimension == 2 || dimension == 3);
+
+  // Using a static map to store meshes by dimension
+  static std::map<int, mesh::PtrMesh> meshes;
+  // Check if the mesh for the given dimension already exists
+  if (meshes.find(dimension) == meshes.end()) {
+    meshes[dimension] = std::make_shared<mesh::Mesh>("(indirect access)", dimension, mesh::Mesh::MESH_ID_UNDEFINED, true);
+  }
+  return meshes[dimension];
+}
```
Why should we and what's wrong with the current approach? There is no value in having potentially multiple indirect access mappings.

davidscn commented on src/mapping/MappingDataCache.hpp:
```diff
@@ -0,0 +1,72 @@
+ * The usual mapping data structures have no notion about the underlying data.
+ * The Cache stores in addition a timestep the data corresponds to.
+ */
+struct MappingDataCache {
```
> This class is currently specific to PUM, right?

For the other methods, we still "need" to cache the time interpolation in the cache. Of course, we can still move it.

davidscn replied:
Moved it to `impl` and added more docs. Reopen if necessary.

davidscn commented on src/mapping/config/MappingConfiguration.cpp:
```diff
@@ -376,6 +376,16 @@ void MappingConfiguration::xmlTagCallback(
     std::string constraint = tag.getStringAttributeValue(ATTR_CONSTRAINT);
 
+    // Check that either of the two is provided
+    PRECICE_CHECK(!toMesh.empty() || !fromMesh.empty(), "Standard Mappings usually need a \"to\" and \"from\" mesh. "
+                                                        "For indirect mesh access, one of both attributes has to be specified.");
```
```suggestion
    PRECICE_CHECK(!toMesh.empty() || !fromMesh.empty(), "Neither a \"to\" nor a \"from\" mesh was defined in a preCICE mapping. Most data mappings require a \"to\" and a \"from\" mesh. "
                                                        "For indirect mesh access, at least one of both attributes has to be specified.");

```

davidscn commented on src/mapping/config/MappingConfiguration.cpp:
```diff
@@ -376,6 +376,16 @@ void MappingConfiguration::xmlTagCallback(
+
+    // Restrict to read-consistent and write-conservative for indirect access
+    PRECICE_CHECK(!toMesh.empty() || (toMesh.empty() && dir == DIRECTION_READ && constraint == CONSTRAINT_CONSISTENT), "For indirect access, only read-consistent (direction = \"read\" and no \"to\" mesh) and write-conservative (direction = \"write\" and no \"from\" mesh) are allowed.");
+    PRECICE_CHECK(!fromMesh.empty() || (fromMesh.empty() && dir == DIRECTION_WRITE && constraint == CONSTRAINT_CONSERVATIVE), "For indirect access, only read-consistent (direction = \"read\" and no \"to\" mesh) and write-conservative (direction = \"write\" and no \"from\" mesh) are allowed.");
```
Done

davidscn commented on src/mapping/NearestNeighborBaseMapping.cpp:
```diff
@@ -109,6 +109,23 @@ void NearestNeighborBaseMapping::tagMeshFirstRound()
   PRECICE_TRACE();
   precice::profiling::Event e("map." + mappingNameShort + ".tagMeshFirstRound.From" + input()->getName() + "To" + output()->getName(), profiling::Synchronize);
 
+  // @TODO: How to handle the parallel partitioning for indirect access here?
+  // in the usual case, we make use of the indexSet, which is pre-computed from the mapping
+  // for the indirect access, we can't do that since we don't have the output mesh
+  // what we would need to do in theory:
+  // find all nearest-neighbors at the 'boundary' of the access region, which would require an
+  // infinite fine sampling of output mesh nodes wo be used in the computeMapping below
+  // for now, we simply tag everything and move on.
+  if (this->isIndirectMapping()) {
+    // Depending on the mapping constraint, one of these tagAll calls will do nothing, as the vertex
+    // set of the mesh is empty. From a practical point of view, we only need to apply the
+    // tagging to one of the meshes (the remote one). But calling it on both sides reliefs us from any
+    // conditional code.
+    output()->tagAll();
+    input()->tagAll();
+    return;
+  }
```
We discussed this in person: the mesh is pre-filtered by the geometric filter, so it's fine to tag everything 👍

davidscn commented on src/mapping/Mapping.hpp:
```diff
@@ -189,6 +194,14 @@ class Mapping {
+  virtual void updateMappingDataCache(MappingDataCache &cache, Eigen::VectorXd &in);
+
+  // For now only for read-consistent
+  virtual void evaluateMappingDataCacheAt(::precice::span<const double> coordinates, const MappingDataCache &cache, ::precice::span<double> values);
```
Done

davidscn commented on src/mapping/PartitionOfUnityMapping.hpp:
```diff
@@ -290,6 +305,93 @@ void PartitionOfUnityMapping<RADIAL_BASIS_FUNCTION_T>::mapConsistent(const time:
+
+  // Step 1: use the indexed center mesh to find relevant cluster
+  query::Index &clusterIndex = _centerMesh->index();
+  // Step 4: find all clusters the output vertex lies in, i.e., find all cluster centers which have the distance of a cluster radius from the given output vertex
```
Fixed

davidscn commented on src/precice/config/ParticipantConfiguration.cpp:
```diff
@@ -438,93 +443,133 @@ void ParticipantConfiguration::finishParticipantConfiguration(
-    const auto &       toMeshID        = confMapping.toMesh->getID();
-    impl::MeshContext &fromMeshContext = participant->meshContext(fromMesh);
-    impl::MeshContext &toMeshContext   = participant->meshContext(toMesh);
+    // The mesh context associated to the mappings
+    // impl::MeshContext &fromMeshContext = ;
+    // impl::MeshContext &toMeshContext   = ;
 
```
Done

davidscn commented on src/precice/Participant.hpp:
```diff
@@ -848,6 +848,88 @@ class PRECICE_API Participant {
+   * This function writes values of temporary vertices to data of a mesh.
+   * As opposed to the writeData function using VertexIDs, this function allows to write data via coordinates,
+   * which don't have to be specified during the initialization. This is particularly useful for meshes, which
+   * vary over time. Note that using this function comes at a performance cost, since the specified mapping
+   * needs to be computed with every call of this function, whereas the other variant can efficiently precompute the mapping.
```
I adjusted the description now slightly. Please check again, I think that the statement is not negative in the first place and also a bit obvious from my perspective.

davidscn commented on src/precice/Participant.hpp:
```diff
@@ -848,6 +848,88 @@ class PRECICE_API Participant {
+   * The 2D-format of values is (d0x, d0y, d1x, d1y, ..., dnx, dny)
+   * The 3D-format of values is (d0x, d0y, d0z, d1x, d1y, d1z, ..., dnx, dny, dnz)
+   *
+   * The time associated to the data is derived from the last call of \ref advance().
```
What's wrong with this sentence? Do you think it is confusing? It might be valuable to get forward this information to the user, that the data is attached to a time stamp, and this is the way we handle this internally in preCICE/the API.

davidscn replied:
Ok, then I remove it.

davidscn commented on src/precice/Participant.hpp:
```diff
@@ -848,6 +848,88 @@ class PRECICE_API Participant {
       double                          relativeReadTime,
       ::precice::span<double>         values) const;
 
+  /**
```
Gave it its own section with some docs.

davidscn commented on tests/parallel/direct-mesh-access/AccessReceivedMeshAndMapping.cpp:
```diff
@@ -9,6 +9,12 @@
 // by another participant (see above). In addition to the direct mesh access
 // and data writing in one direction, an additional mapping (NN) is defined
 // in the other direction.
+// We also test that getMeshVertexSize and getMeshVertexIDsAndCoordinates
+// correctly filter out one vertex (-0.5, -0.5) which is not in the access
+// region, but is used for mapping.
+// Since the user has no opportunity to get the "filtered" vertex (unless
+// the bounding box is enlarged), the read data value on the SolverTwo
+// participant is zero ( last entry in l115 {15, 16, 0}).
```
FYI @uekerman it might happen that vertices are now essentially not visible to the users or filtered out, but I guess that's a logical consequence of how we do now things.

davidscn replied:
The user gets a warning if vertices are filtered out completely during the repartitioning. That's here, however, not the case. We also can't give a warning, as anyway vertices are outside the box due to the halo-region required for the just-in-time mapping.

uekerman replied:
OK, then we did what we could.

uekerman commented on src/precice/Participant.hpp:
```diff
@@ -848,6 +848,88 @@ class PRECICE_API Participant {
+   *
+   * @pre the coordinates are within the bounding box previously defined via \ref setMeshAccessRegion()
+   *
+   * @note Note that the evaluated mapping computes the values corresponding to the initial configuration of the other provided mesh.
```
I could imagine that this formulation is confusing. For preCICE, there is only one configuration, we could call it "initial configuration", but it could also be sth different. For a mesh-particle coupling, the mesh and the particle solver need agree on what to couple. If the mesh solver uses an ALE approach and the particle solver wants to reuse the deformation, you probably have to communicate velocities and mesh deformation. But those are two separate fields and then the particle solver (or the adapter) needs to take to care to use these two the correct way.

Instead of having this sentence here (I would simply remove it), it would be better to eventually have a "6-way" coupled tutorial, we could point to from the docs.

davidscn replied:
I removed it now, though I believe awareness would be good here.

uekerman reviewed (changes requested):
Code looks very clean, but is too enormous to properly review. I could not look at all details of the implementation, but have to trust the tests. Instead, I mainly focused on user-facing things. Tests look very good.

Remarks:
- Shouldn't it be possible to also add unit tests for the new mapping functions?
- We have a dependency on https://github.com/precice/preeco-orga/issues/31 if we want to make calling `setMeshAccessRegion` mandatory.
- Is there an integration test combining JIT with direct access?
- When reading at different points in time and at different locations, is there a performance-wise better order? "First all locations at time one, then all location at time two" or "First time one and two at location one, then time one and two at location two, ..."? If yes, we should properly document as an example in the user docs.

davidscn commented on tests/parallel/just-in-time-mapping/ExplicitRead.cpp:
```diff
@@ -0,0 +1,122 @@
+        } else if (time == 2) {
+          BOOST_TEST(expectedData2[i] == value);
+        } else {
+          PRECICE_ASSERT(false);
```
Yes, but we want to test the correctness of preCICE and not the correctness of the test, right? Semantically, this should be an ASSERT from my understanding. A failing test would indicate malicious behavior in preCICE.

uekerman replied:
But the `time` variable you get indirectly from preCICE, so it would be triggered by malicious behavior in preCICE, no? Probably an edge case. Two disadvantages: `PRECICE_ASSERT` does not get triggered in release builds and leads to a crash in the tests.

davidscn replied:
done

davidscn commented on tests/parallel/just-in-time-mapping/ExplicitRead.cpp:
```diff
@@ -0,0 +1,122 @@
+      } else if (time == 2) {
+        couplingInterface.writeData(meshName, dataName, ids, writeData2);
+      } else {
+        PRECICE_ASSERT(false);
```
See above

davidscn commented on src/precice/Participant.hpp:
The comment was related to the necessity of defining a box. I removed the confusing part.

davidscn commented on src/precice/Participant.hpp:
Fixed

davidscn commented on src/precice/Participant.hpp:
Put it now below the `@brief`

davidscn commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -447,9 +448,20 @@ void ParticipantImpl::advance(
+    // side.
     samplizeWriteData(timeSteppedTo);
   }
+  // TODO: We need to reset here the buffer for just-in-time mappings as well
```
No, was a leftover

davidscn commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -1174,6 +1232,144 @@ void ParticipantImpl::readData(
+{
+  PRECICE_EXPERIMENTAL_API();
+  PRECICE_TRACE(meshName, dataName, coordinates.size(), relativeReadTime);
+  // TODO: Make these checks conditional
```
was outdated

davidscn commented on src/mapping/Mapping.hpp:
```diff
@@ -189,6 +197,53 @@ class Mapping {
+  /**
+   * @brief Allows to update a so-called MappingDataCache for more efficient just-in-time mappings
+   *
+   * To compute a mapping just-in-time (either with \p mapConservativeAt or \p mapConsistentAt ), we
+   * first sample in time and then compute an interpolant in time and then interpolate in space.
```
Yes, this was confusing, rephrased it a bit.

davidscn commented on src/mapping/Mapping.hpp:
```diff
@@ -189,6 +197,53 @@ class Mapping {
+   * @param cache the cache in use
+   * @param in the time-interpolated data values
+   */
+  virtual void updateMappingDataCache(MappingDataCache &cache, Eigen::VectorXd &in);
```
added

davidscn commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -505,6 +527,25 @@ void ParticipantImpl::samplizeWriteData(double time)
   // store buffered write data in sample storage and reset the buffer
   for (auto &context : _accessor->writeDataContexts()) {
+
+    // Finalize conservative write mapping, later we reset
+    // the buffer in resetWrittenData
+    // The function should use _writeDataBuffer internally
+    // context.finishJustInTimeWriteMapping();
```
No, already handled a few lines below.

davidscn commented on src/mapping/Mapping.hpp:
```diff
@@ -189,6 +197,53 @@ class Mapping {
+   * implementations are in derived classes
+   */
+  virtual void mapConsistentAt(const Eigen::Ref<const Eigen::MatrixXd> &coordinates, const MappingDataCache &cache, Eigen::Ref<Eigen::MatrixXd> values);
+  virtual void completeJustInTimeMapping(MappingDataCache &cache, Eigen::Ref<Eigen::MatrixXd> buffer);
```
Added (also in the datacontext)

davidscn commented on src/precice/impl/DataContext.hpp:
```diff
@@ -110,6 +111,42 @@ class DataContext {
   }
 
+  /**
+   * @brief
```
added

davidscn commented on src/precice/impl/WriteDataContext.hpp:
```diff
@@ -71,6 +79,8 @@ class WriteDataContext : public DataContext {
    */
   void appendMappingConfiguration(MappingContext &mappingContext, const MeshContext &meshContext) override;
 
+  void completeJustInTimeMapping();
```
done

davidscn commented:
>     * Shouldn't it be possible to also add unit tests for the new mapping functions?

Added unit tests for the PUM part.

>     * Is there an integration test combining JIT with direct access?
> 

Also added those and checked for the initialization (jit + direct) in various additional places.

>     * When reading at different points in time and at different locations, is there a performance-wise better order? "First all locations at time one, then all location at time two" or "First time one and two at location one, then time one and two at location two, ..."? If yes, we should properly document as an example in the user docs.

Yes. I added a note in the API docs already and will add a more elaborate example in the website docs.

uekerman commented on docs/changelog/2099.md:
```diff
@@ -0,0 +1,4 @@
+- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />`.
```
```suggestion
- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />` and deprecated `direct-access`.
```

uekerman commented on docs/changelog/2099.md:
```diff
@@ -0,0 +1,4 @@
+- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />`.
+- The API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" require now the flag api-access="true". If not set, the function will throw an error.
```
```suggestion
- Fixed the API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" to now require the flag api-access="true". If not set, the function will throw an error.
```
To make it clearer that we do not consider this breaking.

uekerman commented on docs/changelog/2099.md:
```diff
@@ -0,0 +1,4 @@
+- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />`.
+- The API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" require now the flag api-access="true". If not set, the function will throw an error.
+- When combining direct-mesh access with mappings, the function "getMeshVertexIDsAndCoordinates" will return only the vertices within the defined access region ("setMeshAccessRegion"). Setting an access region is mandatory if no other mapping is defined. In parallel it is anyway mandatory.
```
```suggestion
- Changed the behavior of the function "getMeshVertexIDsAndCoordinates", when combining direct-mesh access with mappings, to return only the vertices within the defined access region ("setMeshAccessRegion"). Setting an access region is mandatory if no other mapping is defined. In parallel, it has been mandatory anyway.
```

uekerman commented on docs/changelog/2099.md:
```diff
@@ -0,0 +1,4 @@
+- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />`.
+- The API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" require now the flag api-access="true". If not set, the function will throw an error.
+- When combining direct-mesh access with mappings, the function "getMeshVertexIDsAndCoordinates" will return only the vertices within the defined access region ("setMeshAccessRegion"). Setting an access region is mandatory if no other mapping is defined. In parallel it is anyway mandatory.
+- Added experimental support for an just-in-time mapping with NN and PU-RBF mapping via the new API functions 'mapAndReadData' and 'mapAndWriteData'.
```
```suggestion
- Added experimental support for just-in-time mapping with NN and PU-RBF via the new API functions 'mapAndReadData' and 'writeAndMapData'.
```

BenjaminRodenberg commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -1481,6 +1709,18 @@ void ParticipantImpl::computeMappings(std::vector<MappingContext> &contexts, con
+    }
+  }
+  if (anyMappingChanged) {
+    if (mappingType == "write") {
+      for (auto &context : _accessor->writeDataContexts()) {
+        context.initializeMappingDataCache();
+      }
+    } else {
+      for (auto &context : _accessor->readDataContexts()) {
+        context.initializeMappingDataCache();
+      }
```
Could be factored into `_accessor`. Then call `_accessor.initializeMappingDataCache(mappingType)` here.

BenjaminRodenberg commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -1609,6 +1849,8 @@ void ParticipantImpl::resetWrittenData()
 {
   PRECICE_TRACE();
   for (auto &context : _accessor->writeDataContexts()) {
+    // reset the buffered data here
+    context.invalidateMappingCache(true);
     context.resetBuffer();
```
```suggestion
    // reset the buffered data here
    context.resetBufferedData()
```

and create function inside `WriteDataContext`:

```
void resetBufferedData() {
    invalidateMappingCache(true);
    resetBuffer();  // <- make private or even inline?
}
```

davidscn replied:
put it inline

BenjaminRodenberg commented on src/precice/impl/MeshContext.hpp:
```diff
@@ -35,7 +35,12 @@ struct MeshContext {
 
   /// setMeshAccessRegion may only be called once per mesh(context)
-  bool accessRegionDefined = false;
+  /// putting this into the mesh context means that we can only call
+  /// this once, regardless of combinations with just-in-time mappings
+  /// or multiple such mappings
+  /// if multiples are desired, we would need to shift this into the
+  /// MappingContext
```
Odd line breaking

davidscn replied:
these are individual paragraphs. I can put another empty line in between if this makes it less odd?

BenjaminRodenberg replied:
Why not a multiline comment? You could also add bullet points in the multiline comment to have a better formatting? Right now I really do not understand the rationale behind the line breaks.

davidscn replied:
I removed the line break

BenjaminRodenberg commented on src/precice/impl/DataContext.hpp:
```diff
@@ -110,6 +111,59 @@ class DataContext {
+  void invalidateMappingCache(bool resetData)
+  {
+    if (mappingCache) {
+      mappingCache->setTimeStamp(-1);
```
? The `-1` looks to me like a hack. Why don't define a new function with a better name in the `MappingDataCache`?

```
void MappingDataCache::functionWithDescriptiveName() {
    setTimeStamp(-1);  // add a comment here what -1 means
}
```

BenjaminRodenberg commented on src/precice/config/ParticipantConfiguration.cpp:
```diff
@@ -452,73 +465,111 @@ void ParticipantConfiguration::finishParticipantConfiguration(
-    toMeshContext.toMappingContexts.push_back(mappingContext);
+    // configure the involved mesh context with connectivity requirements stemming from the mapping
+    // Add the mapping context to the mesh context, only required to later on forward them to the Partition
+    if (!input->isJustInTime()) {
+      participant->meshContext(fromMesh).meshRequirement = std::max(
+          participant->meshContext(fromMesh).meshRequirement, map->getInputRequirement());
+      participant->meshContext(fromMesh).fromMappingContexts.push_back(mappingContext);
+    }
+    if (!output->isJustInTime()) {
+      participant->meshContext(toMesh).toMappingContexts.push_back(mappingContext);
+      participant->meshContext(toMesh).meshRequirement = std::max(
+          participant->meshContext(toMesh).meshRequirement, map->getOutputRequirement());
+    }
```
Both `if`-legs could be moved into `MeshContext`. Maybe even the whole block could be moved inside `Mesh`.

davidscn replied:
I am not sure. Then using two functions or what do you have in mind? The two branches are different: input vs output, but also the context we append our own vector to.

BenjaminRodenberg replied:
I mean something in this direction:

```
    if (!input->isJustInTime()) {
      participant->doInputConfiguration(fromMesh, map, mappingContext);
    }
    if (!output->isJustInTime()) {
      participant->doOutputConfiguration(toMesh, map, mappingContext);
    }
```

Maybe some input arguments are not needed. This has two main debefits:

1. the code becomes more readable: you replace the low-level implementation with a high-level function with a descriptive name. 
2. it improves encapsulation: you do not need to access `participant->meshContext(...)` anymore from here but only inside `participant`.

davidscn replied:
Done

BenjaminRodenberg reviewed (changes requested):
I did a quick incomplete review. Two important things:

* `Participant.cpp` and `ParticipantConfiguration.cpp` are already huge and this PR adds a lot of code for both classes. I think there is a lot of potential for refactoring to move code from these two (huge) classes into other classes that are not as complicated.
* Many tests already cover the case `substeps="true"`. However, since we are planning to set `substeps="true"` as the new default it would be important to also test the integration of this feature with the new default and see if any tests break. I'm currently preparing a PR here https://github.com/precice/precice/pull/2220.

davidscn commented:
> However, since we are planning to set substeps="true" as the new default it would be important to also test the integration of this feature with the new default 

I don't understand exactly how this is related to the changes here. As already mentioned, we test here both, substeps="true" and substeps="false". When we change the default in other places, then we either have to explicitly set substeps="false" in failing tests or change the test to test what is supposed to be the correct result. What are you expecting now in this PR?

BenjaminRodenberg commented:
> What are you expecting now in this PR?

Would be interesting to test how the new default and this PR integrates to avoid that suddenly all tests break one day before the release. But I also see this as optional and something that we can check very quickly when either this PR or #2220 is ready to merge. Like you also said: There are plenty of tests that take care about `substeps="true"`.

Much more important is from my perspective the refactoring.

fsimonis commented on src/mapping/RadialBasisFctSolver.hpp:
```diff
@@ -450,6 +467,113 @@ Eigen::VectorXd RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(E
+  if (epsilon.size() > 0) {
+    epsilon -= _matrixQ.transpose() * out;
+    // out  = out - solveTranspose tau (sigma in the PETSc impl)
+    out -= static_cast<Eigen::MatrixXd>(_qrMatrixQ.transpose().solve(-epsilon));
```
Why the static cast?

davidscn replied:
the result of `solve()` (or in general intermediate results in Eigen) is an Eigen internal type. Sometimes, these things are simply views to delay and merge operations defined on those. In this case, the `-=` operator is not defined on the return type of `solve()`. Hence, we explicitly need to cast (i.e., materialize) the result before using the operator.

fsimonis commented on src/mapping/config/MappingConfiguration.cpp:
```diff
@@ -538,6 +553,23 @@ MappingConfiguration::ConfiguredMapping MappingConfiguration::createMapping(
   mesh::PtrMesh     toMesh(_meshConfig->getMesh(toMeshName));
+
+  // Handle for just-in-time mapping, we copy over the dimension and leave everything else
+  if (fromMesh.get() == nullptr && fromMeshName.empty()) {
+    PRECICE_CHECK(toMesh.get() != nullptr,
+                  "Mesh \"{0}\" was not found while creating a mapping. "
+                  "Please correct the to=\"{0}\" attribute.",
+                  toMeshName);
+    fromMesh = mesh::MeshConfiguration::getJustInTimeMappingMesh(toMesh->getDimensions());
+  }
+  if (toMesh.get() == nullptr && toMeshName.empty()) {
+    PRECICE_CHECK(fromMesh.get() != nullptr,
```
This should be equivalent

```suggestion
  if (!fromMesh && fromMeshName.empty()) {
    PRECICE_CHECK(toMesh,
                  "Mesh \"{0}\" was not found while creating a mapping. "
                  "Please correct the to=\"{0}\" attribute.",
                  toMeshName);
    fromMesh = mesh::MeshConfiguration::getJustInTimeMappingMesh(toMesh->getDimensions());
  }
  if (!toMesh && toMeshName.empty()) {
    PRECICE_CHECK(fromMesh,
```

davidscn replied:
Applied it locally to also address the other spots

davidscn commented on extras/bindings/fortran/src/preciceFortran.cpp:
```diff
@@ -434,6 +434,53 @@ try {
+void precicef_write_and_map_data_(
+    const char *meshName,
+    const char *dataName,
+    const int * size,
```
I had the same question when doing this. In its current shape, the design should be coherent with other fortran API functions.

davidscn replied:
Discussion in person: using a pointer is correct

davidscn commented on docs/changelog/2099.md:
```diff
@@ -0,0 +1,4 @@
+- Renamed the 'direct-access' flag in `<receive-mesh direct-access="..." />` to 'api-access' `<receive-mesh api-access="..." />` and deprecated `direct-access`.
+- Fixed the API functions "setMeshAccessRegion" and "getMeshVertexIDsAndCoordinates" to now require the flag api-access="true". If not set, the function will throw an error.
```
Extended the second point

fsimonis commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -685,7 +724,26 @@ int ParticipantImpl::getMeshVertexSize(
+                                                   "but no access region was defined although this is necessary for parallel runs. "
+                                                   "Please define an access region using \"setMeshAccessRegion()\" before calling \"getMeshVertexSize()\".",
+                  meshName);
+    auto result = std::count_if(context.mesh->vertices().cbegin(), context.mesh->vertices().cend(),
+                                [bb = context.userDefinedAccessRegion](const auto &v) { return bb->contains(v); });
```
Let's start moving these to `mesh/Utils.hpp`.

```suggestion
    auto result = mesh::countInBB(context.mesh, context.userDefinedAccessRegion);
```

davidscn commented on src/precice/impl/DataContext.hpp:
```diff
@@ -110,6 +111,59 @@ class DataContext {
+  void initializeMappingDataCache()
+  {
+    if (mappingCache) {
+      justInTimeMapping->initializeMappingDataCache(*mappingCache.get());
```
Didn't remember this somehow

davidscn commented on src/precice/impl/MeshContext.hpp:
```diff
@@ -35,7 +35,12 @@ struct MeshContext {
+  /// or multiple such mappings
+  /// if multiples are desired, we would need to shift this into the
+  /// MappingContext
+  std::shared_ptr<mesh::BoundingBox> userDefinedAccessRegion;
```
Yes, a `unqiue_ptr` deletes the copy constructor, which we need due to shared ownership the the meshcontext itself.

davidscn replied:
using a std::optional would also (hard-) copy the bounding box. I think it would be more safe to keep it as a shared owner ship, even though it is shared among various mesh contexts.

davidscn commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -685,7 +724,26 @@ int ParticipantImpl::getMeshVertexSize(
+                                                   "but no access region was defined although this is necessary for parallel runs. "
+                                                   "Please define an access region using \"setMeshAccessRegion()\" before calling \"getMeshVertexSize()\".",
+                  meshName);
+    auto result = std::count_if(context.mesh->vertices().cbegin(), context.mesh->vertices().cend(),
+                                [bb = context.userDefinedAccessRegion](const auto &v) { return bb->contains(v); });
+    PRECICE_DEBUG("Filtered {} of {} vertices out on mesh {} due to the local access region. Mesh size in the access region: {}", context.mesh->nVertices() - result, context.mesh->nVertices(), meshName, result);
```
moved to mesh utils

davidscn commented on src/mapping/impl/MappingDataCache.hpp:
```diff
@@ -0,0 +1,121 @@
+ * To associate the data the cache holds to a specific timestamp, the cache also stores in
+ * such a timestamp.
+ */
+struct MappingDataCache {
```
contradicts a bit the comment from @BenjaminRodenberg, who mentioned that having descriptive function names for setting times etc improved code readability. I don't have a strong opinion here

davidscn replied:
Moved now to Stamples and all private members are gone. We then need the getters and setters to access the inData stamples, otherwise this is not entirely clear in my opinion

davidscn commented on src/mesh/config/MeshConfiguration.cpp:
```diff
@@ -99,6 +99,19 @@ const PtrDataConfiguration &MeshConfiguration::getDataConfiguration() const
+  PRECICE_ASSERT(dimension == 2 || dimension == 3);
+
+  // Using a static map to store meshes by dimension
+  static std::map<int, mesh::PtrMesh> meshes;
```
Discussion in person: I removed the static declaration: we will have now one mesh per jit mapping

davidscn commented on src/mesh/config/MeshConfiguration.cpp:
```diff
@@ -99,6 +99,19 @@ const PtrDataConfiguration &MeshConfiguration::getDataConfiguration() const
+  // Using a static map to store meshes by dimension
+  static std::map<int, mesh::PtrMesh> meshes;
+  // Check if the mesh for the given dimension already exists
+  if (meshes.find(dimension) == meshes.end()) {
+    meshes[dimension] = std::make_shared<mesh::Mesh>("(just-in-time mapping)", dimension, mesh::Mesh::MESH_ID_UNDEFINED, true);
+  }
+  return meshes[dimension];
```
Obsolete after https://github.com/precice/precice/pull/2099#discussion_r1979676714

davidscn commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -488,6 +500,16 @@ void ParticipantImpl::handleDataAfterAdvance(bool reachedTimeWindowEnd, bool isT
 
+  // Required for implicit coupling
+  for (auto &context : _accessor->readDataContexts()) {
+    context.invalidateMappingCache(false);
```
Done

davidscn commented on src/mapping/NearestNeighborMapping.cpp:
```diff
@@ -26,31 +26,45 @@ NearestNeighborMapping::NearestNeighborMapping(
+  // scalar = 1, vector > 1
+  const int valueDimensions = inData.dataDims;
+  // Create Eigen::Map to interpret data as a matrix (enables col-wise operations)
+  Eigen::Map<const Eigen::MatrixXd> inMap(inData.values.data(), valueDimensions, input()->nVertices());
```
Same as here 

https://github.com/precice/precice/pull/2099#discussion_r1979661095

davidscn commented on src/mapping/impl/MappingDataCache.hpp:
```diff
@@ -0,0 +1,121 @@
+  // interpolation
+  // @note: We could make this a sample rather than a VectorXd, as we might want to operate on gradient data as well
+  // std::unique_ptr<time::Sample> inData;
+  // however, there doesn't seem to be any compatible interface in time::Sample, so we simply use a vector here
```
Moved now to `Stamples` and all private members are gone. We then need the getters and setters to access the `inData` stamples, otherwise this is not entirely clear in my opinion
